### ISSUE TITLE
'no overloads for "sorted" match the provided arguments' for partial when key is an Optional[Callable]

### ISSUE THREAD

**Issue by nytelytee** [microsoft/pyright#9366]
**Describe the bug**
When trying to create a partial for sorted, with the key provided as an optional callable, I get 'no overloads for "sorted" match the provided arguments'.
I looked into typeshed, there are 2 overloads for sorted:
```python
@overload
def sorted(
    iterable: Iterable[SupportsRichComparisonT], /, *, key: None = None, reverse: bool = False
) -> list[SupportsRichComparisonT]: ...
@overload
def sorted(iterable: Iterable[_T], /, *, key: Callable[[_T], SupportsRichComparison], reverse: bool = False) -> list[_T]: ...
```
It seems to me like what I am doing should pass, since the only argument I am providing to partial is a key, which may either be None or a Callable, like the overloads seem to suggest. Mypy reports no error.

**Code or Screenshots**
```python
from functools import partial
from typing import Callable, Optional

# passes fine
def test1(key: Callable):
    return partial(sorted, key=key)

# passes fine
def test2(key: None):
    return partial(sorted, key=key)

# no overloads for "sorted" match the provided arguments
def test3(key: Optional[Callable]):
    return partial(sorted, key=key)
```

**VS Code extension or command-line**
I am running pyright 1.1.387 in the command-line.

**Comment by erictraut** [comment]
Handling of `functools.partial` requires significant special-casing in type checkers. Pyright handles the general case where an overloaded function is passed to `partial`, but it doesn't currently handle the case where argument union expansion is required to match multiple overloads. In this example, there is no single overload signature that accepts a value of type `Callable[..., Any] | None` for the `key` parameter. That means union expansion must be employed to expand the type into its constituent subtypes. 

The typing spec doesn't currently provide any guidance to type checker authors about how overload matching is intended to work. I recently submitted a [new chapter](https://discuss.python.org/t/draft-typing-spec-chapter-for-overloads/60788) that takes a stab at defining the intended behavior. This new section hasn't yet been ratified, and some of the details are likely to change prior to it being finalized. Even once it is finalized, it doesn't say anything about how overload matching should work with `functools.partial`, and that may require additional discussion.

I'll also note that adding support for union expansion for `functools.partial` will require quite a bit of additional special-casing code and carries with it some regression risks. That, combined with the fact that the overload matching specification may force us to revisit the behavior, makes me hesitant to add this functionality at this time.

I'm going to reject the enhancement request for now, but I'm willing to revisit this decision in the future if there is signal from other pyright users that this case is important to them. Hopefully by then the typing spec will have clarified more details about overload matching.
